NIFI-619: Make MonitorActivity more cluster friendly

ijokarumawak opened the pull request:
- Added 'Monitoring Scope' property
- Persist the latest activity as Cluster state
- Examine cluster state on each node if necessary

<img width="762" alt="monitoractivity-config" src="https://cloud.githubusercontent.com/assets/1107620/16346160/d7526104-3a80-11e6-85eb-f9a6fd8af206.png">

## Test

- Setup data flow as shown in the screenshot, using a 3 node cluster
- GenerateFlowFile is scheduled on primary.
- Monitoring Scope = Node
     - [ok] Even if GenerateFlowFile is running, non-primary nodes will report inactive messages.
     - [ok] If GenerateFlowFile stops, all nodes report inactive messages.
     - [ok] If GenerateFlowFile resumes, primary node reports activity.restored, but other nodes don't
- Monitoring Scope = Cluster && Reporting = all
     - [ok] While GenerateFlowFile is running, none of nodes reports inactive messages.
     - [ok] If GenerateFlowFile stops, all nodes report inactive messages.
     - [ok] If GenerateFlowFile resumes, all nodes report activity.restored, if copy attribute is true, then non-primary nodes copy attributes from state.
- Monitoring Scope = Cluster && Reporting = primary
     - [ok] While GenerateFlowFile is running, none of nodes reports inactive messages.
     - [ok] If GenerateFlowFile stops, primary node reports inactive messages. Others don't.
     - [ok] If GenerateFlowFile resumes, primary node reports activity.restored. Others don't.
     - [ok] Disconnect primary node and other node takes over primary node.
     - [ok] Login to the disconnected node, if GenerateFlowFile stops, inactive message should be reported, even if MonitorActivity is using cluster & primary, fallback to standalone mode
     - [ok] And when it resumes, activity restored message should be reported, even if MonitorActivity is using cluster & primary, fallback to standalone mode
- [ok] Confirmed it works on Standalone NiFi instance, too

<img width="955" alt="monitoractivity-test" src="https://cloud.githubusercontent.com/assets/1107620/16346162/d920d344-3a80-11e6-886b-2acd54315968.png">

JPercivall commented:
Hey @ijokarumawak,

I'm gonna be reviewing this today but did you check out the Travis build? It is failing on a MonitorActivity Test:

Tests run: 7, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 9.555 sec <<< FAILURE! - in org.apache.nifi.processors.standard.TestMonitorActivity
testClusterMonitorActivityRestoredBySelf(org.apache.nifi.processors.standard.TestMonitorActivity)  Time elapsed: 0.015 sec  <<< FAILURE!
java.lang.AssertionError: Latest timestamp should be persisted
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertNotNull(Assert.java:712)
	at org.apache.nifi.processors.standard.TestMonitorActivity.testClusterMonitorActivityRestoredBySelf(TestMonitorActivity.java:323)

JPercivall commented:
When running clustered and monitoring on the cluster, it needs to take into account generating the inactive flowfiles[1] in a cluster friendly way as well. Currently each Node will generate a flowfile when the inactivity time hits. Probably need to track in clustered state whether or not a marker has been sent.


[1] https://github.com/apache/nifi/blob/3726e34eb6cb5c568ee14fa0262faf1968435dd8/nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/MonitorActivity.java#L233-L233

JPercivall commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/MonitorActivity.java:
```diff
@@ -201,17 +253,57 @@ public void process(final OutputStream out) throws IOException {
+                    } else {
+                        // If this returns false due to race condition, it's not a problem since we just need
+                        // the latest active timestamp.
+                        stateManager.replace(state, newValues, Scope.CLUSTER);
```
While it usually is the case, we can't assume that the new state has a later timestamp. It could be case that one node is much more taxed than another and it takes longer to transmit state back to the ZooKeeper cluster (where clustered state is stored). This should continually check/attempt to replace until it succeeds or the stored state is later than the current timestamp.

ijokarumawak replied:
Thanks for catching this, I'll add code to check if the current timestamp is newer than the one stored in Zookeeper.

For retrying, I'd let NiFi scheduler and onTrigger() manage that, instead of adding a retry mechanism here, for simplicity and controllability. If it's trying to update state, it means that the node is active, so it will report correctly. The only problem I can imagine from not updating the state, is the possibility for other nodes to emit false alarms. But it's also inevitable if other nodes themselves are inactive and Zk is not working. If the issue is solved, further onTrigger() call will write state and it will work correctly eventually.

JPercivall commented:
Also you need to double check that the generate activity resumed flowfile notifications are done in a cluster friendly way too (once per resumption of activity).

ijokarumawak commented:
@JPercivall Thanks for reviewing this.

For the 1st comment:

I didn't notice that it's failing Travis test... I'll check what caused the error. Thanks for pointing that.

For the 2nd and last comment: 

> When running clustered and monitoring on the cluster, it needs to take into account generating the inactive flowfiles in a cluster friendly way as well. Currently each Node will generate a flowfile when the inactivity time hits. Probably need to track in clustered state whether or not a marker has been sent.

I wrote onTrigger like that intentionally, because I thought there're use-cases in which user would like to do something on each node when NiFi went inactive or recovered activity. If one would like to get notification flow-file only on a single node in the cluster, then they can use `On primary node` scheduling strategy. Excuse me if I misunderstood your comments, but does this address your concern?

ijokarumawak commented:
@JPercivall I was wrong with above comment. If one uses `On primary node`, then its onTrigger won't be executed, it means it can't update state in Zookeeper either. I'll add property to indicate whether inactive / restored notification should be sent only from primary node.

ijokarumawak commented:
@JPercivall I've addressed review comments, pushed another commit so that you can see difference from the last one.
Confirmed that unit tests passed on Travis.

Would you take a look once again? Thank you!

JPercivall commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/MonitorActivity.java:
```diff
@@ -168,17 +216,49 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
+        boolean isInactive = false;
+        long updatedLatestSuccessTransfer = -1;
+        StateMap clusterState = null;
+        final boolean shouldThisNodeReport = !isClusterScope
```
I like the change to add the ability to report on the primary node or all. That said, this check should probably be moved to a function that is used before each time it would report. Since the primary node change could happen at any time this function should wait to do the check to try and prevent two nodes from each sending a message.

ijokarumawak replied:
Good call. I'll move this condition to a function and call it just before reporting attempts.

JPercivall commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/MonitorActivity.java:
```diff
@@ -234,4 +361,10 @@ public void process(final OutputStream out) throws IOException {
+
+    @OnPrimaryNodeStateChange
+    public void onPrimaryNodeChange(final PrimaryNodeState newState) {
+        isPrimaryNode = (newState == PrimaryNodeState.ELECTED_PRIMARY_NODE);
```
I am running a secure cluster and when I have "primary node only" configured for the sending of the messages they never get sent. I believe this is because the primary node has been the same since the processor was added and it was never changed, therefore this method would never have gotten called. 

I think this needs some way to know that the processor is running on the primary node when it gets scheduled or initialized. I'm not sure if that is possible at the moment though.

ijokarumawak replied:
This is a great catch, thank you!

@JPercivall in different PR I'm working on, #563 also needs similar functionality to know whether it's running on a standalone or clustered NiFi, and if latter whether it's a primary node.

I'm going to try adding a new `@OnAddedToCluster` lifecycle annotation taking one argument `isPrimaryNode`. Using it together with `@OnPrimaryNodeStateChange` Processors, Reporting tasks, and Controller services will be able to control its behavior well in a clustered environment.

How do you think?

JPercivall replied:
(adding for transparency) @ijokarumawak and I talked out of band. Current thinking is to add another variable to the Initialization Contexts in order to know when the component is created if it is clustered or not and if it is on the primary node.

ijokarumawak commented:
@JPercivall I've updated the PR:

- Rebased and squashed
- Added NodeTypeProvider to expose flowController's isClustered and  isPrimaryNode so that processor can know if it's running on a cluster and if it's a primary node
- Removed `@OnPrimaryNodeStateChange` because NodeTypeProvider checks current state so no longer need to be notified
- Updated mock framework accordingly to support cluster/primary node simulation
- Added Property check regarding Node type

During test, I found a bug and fixed it, too. That is, after MonitorActivity got `inactive` mode,  stop the processor and restart the processor, then it emits `activity.restored` flow-files immediately. Because the instance hold `AtomicBoolean inactive` with previous state. I've added a code to reset it with default value when `@OnScheduled` is called.

Also, added `@OnStopped` to clear persisted state, so that further activity check won't get affected by the stale state.

The change had been bigger than I originally thought... but NodeTypeProvider covers what I was trying to accomplish. Please take a look. Thanks for your support!

JPercivall commented:
@ijokarumawak the work adding new functionality to determine if any processor/component is running clustered or is the primary node should be a part of a separate issue (ticket) and PR. It is outside the scope of the ticket "Make MonitorActivity more cluster friendly"

JPercivall commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/FlowController.java:
```diff
@@ -1004,7 +1004,7 @@ public ProcessorNode createProcessor(final String type, String id, final boolean
         }
 
-        final ValidationContextFactory validationContextFactory = new StandardValidationContextFactory(controllerServiceProvider);
+        final ValidationContextFactory validationContextFactory = new StandardValidationContextFactory(controllerServiceProvider, this);
```
I really don't like passing the FlowController itself. Why can't it just pass the value? Or if it needs to be able to be updated, it could pass a shared reference.

ijokarumawak replied:
I don't like it, too. So I passed it as NodeTypeProvider, only isClustered and isPrimaryNode are accessible.

ijokarumawak replied:
But yes, if one tries to cast it as FlowController, it's possible..

JPercivall commented on nifi-api/src/main/java/org/apache/nifi/components/ValidationContext.java:
```diff
@@ -99,4 +100,9 @@
+    /**
+     * @return the {@link NodeTypeProvider} which can be used to detect the node type of this NiFi instance.
+     */
+    NodeTypeProvider getNodeTypeProvider();
```
I don't think this should be exposed here. The primary node check is inherently dangerous for a validation check and something we shouldn't encourage. Instead let the user get the state in the init and use the clustered value in the validation check.

ijokarumawak replied:
Agreed. I was thinking it'd be too much. Exposing primary state was not what I wanted, though.. Processor's property descriptors and validators are often defined in static scope, and I thought if ValidationContext provides cluster or not, it's useful.

I'll remove this.

ijokarumawak commented:
@JPercivall I'm running out of time today.
Also, I realized isClustered can change after processor is created. If a node is disconnected from a cluster, it will be a standalone.

I'll reconsider how to proceed this, thanks.

JPercivall commented:
Hey @ijokarumawak, I will review the NIFI-2267 changes on the other PR. Once that is finished we can resume work on this one (since it is blocked by that 2267).

JPercivall commented:
+1

Verified the code and any comments were addressed, did a contrib check build and ran through a bunch of tests using a standalone instance and secure cluster, all ran as expected. Looks good, thanks @ijokarumawak

ijokarumawak commented:
@JPercivall Thanks for the great review process!
